src/base/zonefile_fmt.rs

```diff
@@ -0,0 +1,346 @@
+}
+
+impl<'a> SimpleWriter<'a> {
+    fn new(writer: &'a mut dyn fmt::Write) -> Self {
+        Self {
+            first: true,
+            writer,
+        }
+    }
+}
+
+impl PresentationWriter for SimpleWriter<'_> {
+    fn fmt_token(&mut self, args: fmt::Arguments<'_>) -> Result {
+        if !self.first {
+            self.writer.write_char(' ')?;
+        }
+        self.first = false;
+        self.writer.write_fmt(args)?;
+        Ok(())
+    }
+
+    fn begin_block(&mut self) -> Result {
+        Ok(())
+    }
+
+    fn end_block(&mut self) -> Result {
+        Ok(())
+    }
+
+    fn fmt_comment(&mut self, _args: fmt::Arguments<'_>) -> Result {
+        Ok(())
+    }
+
+    fn newline(&mut self) -> Result {
+        self.writer.write_char('\n')?;
+        self.first = true;
+        Ok(())
+    }
+}
+
+struct MultiLineWriter<'a> {
+    current_column: usize,
+    block_indent: Option<usize>,
+    first: bool,
+    writer: &'a mut (dyn fmt::Write + 'a),
+}
+
+impl<'a> MultiLineWriter<'a> {
+    fn new(writer: &'a mut dyn fmt::Write) -> Self {
+        Self {
+            first: true,
+            current_column: 0,
+            block_indent: None,
+            writer,
+        }
+    }
+}
+
+impl PresentationWriter for MultiLineWriter<'_> {
+    fn fmt_token(&mut self, args: fmt::Arguments<'_>) -> Result {
+        use fmt::Write;
+        if !self.first {
+            self.write_str(" ")?;
+        }
+        self.first = false;
+        self.write_fmt(args)?;
+        Ok(())
+    }
+
+    fn begin_block(&mut self) -> Result {
+        self.fmt_token(format_args!("("))?;
+        self.block_indent = Some(self.current_column + 1);
+        Ok(())
+    }
+
+    fn end_block(&mut self) -> Result {
+        self.block_indent = None;
+        self.fmt_token(format_args!(")"))
+    }
+
+    fn fmt_comment(&mut self, args: fmt::Arguments<'_>) -> Result {
+        if self.block_indent.is_some() {
+            self.writer.write_fmt(format_args!("\t; {}", args))?;
+            self.newline()
+        } else {
+            // a comment should not have been allowed
+            // so ignore it
+            Ok(())
+        }
+    }
+
+    fn newline(&mut self) -> Result {
+        use fmt::Write;
+        self.writer.write_char('\n')?;
+        self.current_column = 0;
+        if let Some(x) = self.block_indent {
+            for _ in 0..x {
+                self.write_str(" ")?;
+            }
+        }
+        self.first = true;
+        Ok(())
+    }
+}
+
+impl fmt::Write for MultiLineWriter<'_> {
+    fn write_str(&mut self, x: &str) -> fmt::Result {
+        self.current_column += x.len();
+        self.writer.write_str(x)
+    }
+}
+
+/// A more structured wrapper around a [`PresentationWriter`]
+///
+/// Writing comments is not allowed with this type because comments can only
+/// appear when a token is surrounded by parentheses.
+pub struct Presenter<'a> {
+    writer: &'a mut (dyn PresentationWriter + 'a),
```

Comment:
This shouldn't be a trait object.  `std::fmt` uses this paradigm internally and they've suffered from a lot of performance issues because of it.  Based on the uses I've seen in this crate, I think `Presenter` can be generic over the underlying writer type, or the functions it provides can be moved to `PresentationWriter` instead with default impls.  Changing this interface later will be a lot harder, and I think we should get ahead of this while we can.

Reply:
Fiiiiiiiiiiine, but only because Philip cared about performance too :smile: